kaggle_environments/envs/halite/helpers.py

```diff
@@ -0,0 +1,781 @@
+                ship_position = index_to_position(ship_index, size)
+                raw_action = player_actions.get(ship_id)
+                action = (
+                    ShipAction[raw_action]
+                    if raw_action is not None
+                    else None
+                )
+                self._add_ship(Ship(ship_id, ship_position, ship_halite, player_id, self, action))
+
+            for (shipyard_id, shipyard_index) in player_shipyards.items():
+                shipyard_position = index_to_position(shipyard_index, size)
+                raw_action = player_actions.get(shipyard_id)
+                action = (
+                    ShipyardAction[raw_action]
+                    if raw_action is not None
+                    else None
+                )
+                self._add_shipyard(Shipyard(shipyard_id, shipyard_position, player_id, self, action))
+
+    @property
+    def configuration(self) -> Configuration:
+        return self._configuration
+
+    @property
+    def players(self) -> Dict[PlayerId, Player]:
+        return self._players
+
+    @property
+    def ships(self) -> Dict[ShipId, Ship]:
+        """Returns all ships on the current board."""
+        return self._ships
+
+    @property
+    def shipyards(self) -> Dict[ShipyardId, Shipyard]:
+        """Returns all shipyards on the current board."""
+        return self._shipyards
+
+    @property
+    def cells(self) -> Dict[Point, Cell]:
+        """Returns all cells on the current board."""
+        return self._cells
+
+    @property
+    def step(self) -> int:
+        return self._step
+
+    @property
+    def current_player_id(self) -> PlayerId:
+        return self._current_player_id
+
+    @property
+    def current_player(self) -> Player:
+        """Returns the current player (generally this is you)."""
+        return self._players[self.current_player_id]
+
+    @property
+    def opponents(self) -> List[Player]:
+        """
+        Returns all players that aren't the current player.
+        You can get all opponent ships with [ship for ship in player.ships for player in board.opponents]
+        """
+        return [player for player in self.players.values() if not player.is_current_player]
+
+    @property
+    def observation(self) -> Dict[str, Any]:
+        """Converts a Board back to the normalized observation that constructed it."""
+        size = self.configuration.size
+        halite = [self[index_to_position(index, size)].halite for index in range(size * size)]
+        players = [player.observation for player in self.players.values()]
+
+        return {
+            "halite": halite,
+            "players": players,
+            "player": self.current_player_id,
+            "step": self.step
+        }
+
+    def __deepcopy__(self, _) -> 'Board':
+        actions = [player.next_actions for player in self.players.values()]
+        return Board(self.observation, self.configuration, actions)
```

Comment:
Because the transformation to an observation copies all of the data into new data structures we can use it for deepcopy as well to isolate a copied instance from its parent.

Reply:
`next_actions` is a computed property constructed from the player's ships and shipyards so we don't have to worry about modification